core/src/main/scala/org/apache/spark/scheduler/BlacklistTracker.scala

```diff
@@ -0,0 +1,393 @@
+    // they will grow too big since you won't get too many executors on one node, and the timeout
+    // will clear it up periodically in any case.
+    executorIdToFailureList -= executorId
+  }
+}
+
+
+private[scheduler] object BlacklistTracker extends Logging {
+
+  private val DEFAULT_TIMEOUT = "1h"
+
+  /**
+   * Returns true if the blacklist is enabled, based on checking the configuration in the following
+   * order:
+   * 1. Is it specifically enabled or disabled?
+   * 2. Is it enabled via the legacy timeout conf?
+   * 3. Use the default for the spark-master:
+   *   - off for local mode
+   *   - on for distributed modes (including local-cluster)
+   */
+  def isBlacklistEnabled(conf: SparkConf): Boolean = {
+    conf.get(config.BLACKLIST_ENABLED) match {
+      case Some(isEnabled) =>
+        isEnabled
+      case None =>
+        // if they've got a non-zero setting for the legacy conf, always enable the blacklist,
+        // otherwise, use the default based on the cluster-mode (off for local-mode, on otherwise).
+        val legacyKey = config.BLACKLIST_LEGACY_TIMEOUT_CONF.key
+        conf.get(config.BLACKLIST_LEGACY_TIMEOUT_CONF) match {
+          case Some(legacyTimeout) =>
+            if (legacyTimeout == 0) {
+              logWarning(s"Turning off blacklisting due to legacy configuaration:" +
+                s" $legacyKey == 0")
+              false
+            } else {
+              // mostly this is necessary just for tests, since real users that want the blacklist
+              // will get it anyway by default
+              logWarning(s"Turning on blacklisting due to legacy configuration:" +
+                s" $legacyKey > 0")
+              true
+            }
+          case None =>
+            // local-cluster is *not* considered local for these purposes, we still want the
+            // blacklist enabled by default
+            !Utils.isLocalMaster(conf)
+        }
+    }
+  }
+
+  def getBlacklistTimeout(conf: SparkConf): Long = {
+    conf.get(config.BLACKLIST_TIMEOUT_CONF).getOrElse {
+      conf.get(config.BLACKLIST_LEGACY_TIMEOUT_CONF).getOrElse {
+        Utils.timeStringAsMs(DEFAULT_TIMEOUT)
+      }
+    }
+  }
+
+  /**
+   * Verify that blacklist configurations are consistent; if not, throw an exception.  Should only
+   * be called if blacklisting is enabled.
+   *
+   * The configuration for the blacklist is expected to adhere to a few invariants.  Default
+   * values follow these rules of course, but users may unwittingly change one configuration
+   * without making the corresponding adjustment elsewhere.  This ensures we fail-fast when
+   * there are such misconfigurations.
+   */
+  def validateBlacklistConfs(conf: SparkConf): Unit = {
+
+    def mustBePos(k: String, v: String): Unit = {
+      throw new IllegalArgumentException(s"$k was $v, but must be > 0.")
+    }
+
+    // undocumented escape hatch for validation -- just for tests that want to run in an "unsafe"
+    // configuration.
```

Comment:
yeah, thats a fair point.  Its definitely not needed for checking correct behavior -- its only for this "test": https://github.com/squito/spark/blob/blacklist-SPARK-8425/core/src/test/scala/org/apache/spark/scheduler/BlacklistIntegrationSuite.scala#L71

The idea of that test is more to document / demonstrate behavior of the system for a developer that is trying to understand how the system works & why its designed as it is, rather than test for correct behavior (except perhaps they help make sure the tests are testing what they are supposed to).   I find this kind of thing useful both as I'm developing, and as I'm trying to understand other parts of the code even if I'm not adding features.

I don't think there is a huge risk to a user accidentally stumbling on this, its undocumented, and if they find it by browsing code they clearly see its not meant for the end-user.

But spark itself doesn't seem to have many "tests" like the one I'm adding, so I understand if you think we should remove that as well.